src/OrchardCore.Modules/OrchardCore.Queries/Services/QueriesDocumentUpdater.cs

```diff
@@ -0,0 +1,77 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Text.Json.Nodes;
+using System.Threading.Tasks;
+using OrchardCore.Documents;
+using OrchardCore.Environment.Extensions;
+using OrchardCore.Environment.Extensions.Features;
+using OrchardCore.Environment.Shell;
+
+namespace OrchardCore.Queries.Services;
+
+public sealed class QueriesDocumentUpdater : FeatureEventHandler
+{
+    private readonly IDocumentManager<QueriesDocument> _documentManager;
+    private readonly IEnumerable<IQuerySource> _querySources;
+    private readonly ITypeFeatureProvider _typeFeatureProvider;
+
+    public QueriesDocumentUpdater(
+        IDocumentManager<QueriesDocument> documentManager,
+        IEnumerable<IQuerySource> querySources,
+        ITypeFeatureProvider typeFeatureProvider)
+    {
+        _documentManager = documentManager;
+        _querySources = querySources;
+        _typeFeatureProvider = typeFeatureProvider;
+    }
+
+    public override async Task EnabledAsync(IFeatureInfo feature)
+    {
+        // If the newly activated feature contains a IQuerySource, update the QueriesDocument in case that source
+        // type has been used before.
+        if (_querySources.Any(source => _typeFeatureProvider.GetFeaturesForDependency(source.GetType()).Any(f => f.Id == feature.Id)))
+        {
+            await FixupDocumentAsync();
+        }
+    }
+
+    private async Task FixupDocumentAsync()
+    {
+        var document = await _documentManager.GetOrCreateMutableAsync();
+
+        if (!ValidateDocument(document) && FixupDocument(document))
```

Comment:
I think `!ValidateDocument(document)` is unnecessary should be removed. I think using `FixupDocument` will be enough since we'll only iterate once instead of possible two. Also maybe call this method `RepairDocument` instead.